components/webext-storage/src/sync/incoming.rs

```diff
@@ -0,0 +1,470 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+// The "incoming" part of syncing - handling the incoming rows, staging them,
+// working out a plan for them, updating the local data and mirror, etc.
+
+use interrupt::Interruptee;
+use rusqlite::{types::ToSql, Connection, Row, Transaction};
+use serde_json;
+use sql_support::ConnExt;
+use sync_guid::Guid as SyncGuid;
+
+use crate::error::*;
+
+use super::{merge, JsonMap, ServerPayload};
+
+// This module deals exclusively with the Map inside a JsonValue::Object().
+// This helper reads such a Map from a SQL row, ignoring anything which is
+// either invalid JSON or a different JSON type.
+fn json_map_from_row(row: &Row<'_>, col: &str) -> Result<Option<JsonMap>> {
+    let s = row.get::<_, Option<String>>(col)?;
+    Ok(match s {
+        None => None,
+        Some(s) => match serde_json::from_str(&s) {
+            Ok(serde_json::Value::Object(m)) => Some(m),
+            _ => {
+                // We don't want invalid json or wrong types to kill syncing -
+                // but it should be impossible as we never write anything which
+                // could cause it, so logging shouldn't hurt.
+                log::warn!("skipping invalid json in {}", col);
+                None
+            }
+        },
+    })
+}
+
+/// The first thing we do with incoming items is to "stage" them in a temp table.
+/// The actual processing is done via this table.
+pub fn stage_incoming<S: ?Sized + Interruptee>(
+    tx: &Transaction<'_>,
+    incoming_bsos: Vec<ServerPayload>,
+    signal: &S,
+) -> Result<()> {
+    // markh always struggles with the sql_support chunking :( So take the
+    // low road...
+    let sql = "
+        INSERT OR REPLACE INTO temp.storage_sync_staging
+        (guid, ext_id, data, server_modified)
+        VALUES (:guid, :ext_id, :data, :ts)";
+    for bso in incoming_bsos {
+        signal.err_if_interrupted()?;
+        tx.execute_named_cached(
+            &sql,
+            &[
+                (":guid", &bso.guid as &dyn ToSql),
+                (":ext_id", &bso.ext_id),
+                (":data", &bso.data),
+                (":ts", &bso.last_modified.as_millis()),
+            ],
+        )?;
+    }
+    Ok(())
+}
+
+/// Details about an incoming item.
+#[derive(Debug, PartialEq)]
+pub struct IncomingItem {
+    guid: SyncGuid,
+    ext_id: String,
+}
+
+/// The "state" we find ourselves in when considering an incoming/staging
+/// record. This "state" is the input to calculating the IncomingAction.
+#[derive(Debug, PartialEq)]
+pub enum IncomingState {
+    IncomingOnly {
```

Comment:
I'd also like to express that the types themselves are confusing to me. I would have assumed that e.g. `IncomingOnly` would require a non-`Option` `JsonMap`. The fact that all of these are `Options` in all variants makes me wonder what the difference is between a simple `struct IncomingState { incoming: Option<JsonMap>, local: Option<JsonMap>, mirror: Option<JsonMap> }`. I'm guessing that the use of `Option` here doesn't necessarily mean "missing" but means something more like tombstones?

Reply:
yes, exactly, it means a tombstone. I can probably make that more explicit by using yet another enum to make it more explicit. I agree the distinction between an empty object and an explicit tombstone isn't clear either.

TBH though, I've been kinda deferring this until this discussion about the semantics.  IIUC, if we went with Chrome's observed behaviour, we wouldn't even need tombstones, but because we are cloning the existing impl, we do, right?

How does the current impl work in this regard? Is there any distinction between an explicit deletion vs an empty object? But as above, I'm not tied to anything here in any way, as I was just ignoring it :)

Reply:
My understanding of Chrome's behavior is that the client has tombstones but the server doesn't. So I think we would still need to represent tombstones to *some* extent. The current implementation does have tombstones both on the client and on the server. Actually, the current implementation has tombstones on a per-key basis, which I don't think we can easily represent given the structures we have now, but maybe I'm missing something.

Reply:
I ended up adding another Enum -

```
/// The state data can be in. Could be represented as Option<JsonMap>, but this
/// is clearer and independent of how the data is stored.
#[derive(Debug, PartialEq)]
pub enum DataState {
    /// The data was deleted.
    Deleted,
    /// Data exists, as stored in the map.
    Exists(JsonMap),
}
```

Which is more verbose, but does make the various states easier to read. I pushed it in rev e478c2142248ad89cf638d0b35d326a45b8101aa - so it's easy to see the changes, but also easy to back out if anyone things it's a backwards step - LMKWYT!